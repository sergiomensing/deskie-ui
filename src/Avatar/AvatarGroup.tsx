import { HTMLAttributes, PropsWithChildren, Children } from 'react';
import styled, { css } from 'styled-components';

export type AvatarGroupProps = HTMLAttributes<unknown> & {
  count?: number;
  direction?: 'horizontal' | 'vertical';
};

const StyledAvatarGroup = styled.div<{ horizontal: boolean }>`
  display: flex;
  ${(props) =>
    props.horizontal
      ? css`
          flex-direction: row;
          & > * + * {
            margin-left: 0.5rem;
          }
        `
      : css`
          flex-direction: column;
          & > * + * {
            margin-top: 0.5rem;
          }
        `};

  .avatar-img {
    width: 3rem;
    height: 3rem;
    border-radius: 1rem;
  }

  .avatar-initials {
    width: 3rem;
    height: 3rem;
    border-radius: 1rem;
  }

  .avatarGroup-additional {
    width: 3rem;
    height: 3rem;
    border-radius: 1rem;
    background: ${(props) => props.theme.colors.accent1};
    display: flex;
    justify-content: center;
    align-items: center;
    font-size: 0.875rem;
    color: ${(props) => props.theme.colors.foreground};
    font-weight: 600;
  }
`;

function AvatarGroup({ children, count, direction = 'horizontal' }: PropsWithChildren<AvatarGroupProps>) {
  function renderLimitedChildren() {
    if (count) {
      const totalAmount = Children.toArray(children);
      const amountNotShown = totalAmount.length - count;
      return (
        <>
          {totalAmount.map((child, index) => {
            return index < count ? child : null;
          })}
          <span className="avatarGroup-additional">+{amountNotShown}</span>
        </>
      );
    }

    return children;
  }

  return (
    <StyledAvatarGroup horizontal={direction === 'horizontal'} className="avatarGroup">
      {renderLimitedChildren()}
    </StyledAvatarGroup>
  );
}

export default AvatarGroup;
