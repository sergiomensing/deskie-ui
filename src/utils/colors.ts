export const colorShade = (col: string, amt: number) => {
  col = col.replace(/^#/, '');
  if (col.length === 3) col = col[0] + col[0] + col[1] + col[1] + col[2] + col[2];

  const [sr, sg, sb] = col.match(/.{2}/g) || ['ff', 'ff', 'ff'];
  const [ar, ag, ab] = [parseInt(sr, 16) + amt, parseInt(sg, 16) + amt, parseInt(sb, 16) + amt];

  const r = Math.max(Math.min(255, ar), 0).toString(16);
  const g = Math.max(Math.min(255, ag), 0).toString(16);
  const b = Math.max(Math.min(255, ab), 0).toString(16);

  const rr = (r.length < 2 ? '0' : '') + r;
  const gg = (g.length < 2 ? '0' : '') + g;
  const bb = (b.length < 2 ? '0' : '') + b;

  return `#${rr}${gg}${bb}`;
};

const defaultColors = [
  '#E64980',
  '#4C6EF5',
  '#228BE6',
  '#15AABF',
  '#12B886',
  '#40C057',
  '#82C91E',
  '#FAB005',
  '#FD7E14',
  '#FA5252',
];

export function getColorFromString(value: string, colors: undefined | string[] = defaultColors) {
  let hash = 0;
  if (value.length === 0) return colors[hash];
  for (let i = 0; i < value.length; i++) {
    hash = value.charCodeAt(i) + ((hash << 5) - hash);
    hash = hash & hash;
  }
  hash = ((hash % colors.length) + colors.length) % colors.length;
  return colors[hash];
}
