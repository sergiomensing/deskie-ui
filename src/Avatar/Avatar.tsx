import { HTMLAttributes } from 'react';
import styled, { css } from 'styled-components';
import initials from 'initials';
import { getColorFromString } from '../utils/colors';
import AvatarGroup from './AvatarGroup';

type AvatarSize = 'default' | 'large';

export type AvatarProps = HTMLAttributes<unknown> & {
  text: string;
  showLabel?: boolean;
  subText?: string;
  size?: AvatarSize | number;
  src?: string;
  isCurrentUser?: boolean;
};

const getSize = (size: AvatarSize | number): string => {
  const sizes = {
    default: '3rem',
    large: '6rem',
  };

  if (typeof size === 'number') return `${size}px`;
  return sizes[size];
};

const getBorderRadius = (size: AvatarSize | number): string => {
  const sizes = {
    default: '1rem',
    large: '2rem',
  };

  if (typeof size === 'number') return `${size / 4}px`;
  return sizes[size];
};

const StyledAvatar = styled.div<{ size: string; borderRadius: string; isCurrentUser: boolean }>`
  display: flex;
  align-items: center;

  & > *:first-child {
    width: ${(props) => props.size};
    height: ${(props) => props.size};
    border-radius: ${(props) => props.borderRadius};
    display: flex;
    user-select: none;
  }

  .avatar-initials {
    color: ${(props) => props.theme.colors.accent1};
    justify-content: center;
    align-items: center;
    text-align: center;

    ${(props) =>
      props.isCurrentUser &&
      css`
        background-color: ${(props) => props.theme.colors.primary} !important;
      `};
  }

  .avatar-label {
    margin-left: 0.75rem;
    display: flex;
    flex-direction: column;
  }

  .avatar-text {
    font-size: 0.875rem;
    color: ${(props) => (props.isCurrentUser ? props.theme.colors.primary : props.theme.colors.foreground)};
    font-weight: 500;
  }

  .avatar-subtext {
    font-size: 0.8125rem;
    color: ${(props) => props.theme.colors.accent8};
    margin-top: 0.25rem;
  }
`;

function Avatar({
  text,
  src,
  showLabel = false,
  subText,
  size = 'default',
  className = '',
  style,
  isCurrentUser = false,
  ...rest
}: AvatarProps) {
  return (
    <StyledAvatar
      className={`avatar ${className}`}
      size={getSize(size)}
      borderRadius={getBorderRadius(size)}
      isCurrentUser={isCurrentUser}
      style={style}
      {...rest}
    >
      {src ? (
        <img className="avatar-img" src={src} alt={text} />
      ) : (
        <div className="avatar-initials" style={{ backgroundColor: getColorFromString(text) }}>
          {initials(text)}
        </div>
      )}
      {showLabel && (
        <div className="avatar-label">
          <span className="avatar-text">{text}</span>
          {subText && <span className="avatar-subtext">{subText}</span>}
        </div>
      )}
    </StyledAvatar>
  );
}

Avatar.Group = AvatarGroup;

export default Avatar;
