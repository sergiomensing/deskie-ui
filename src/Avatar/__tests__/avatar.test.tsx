import React from 'react';
import Avatar from '../Avatar';
import { render } from '../../../tests/utils';

describe('Avatar', () => {
  it('should render with image when available', () => {
    const { container } = render(<Avatar text="Test Value" src="/img.jpg" />);
    expect(container.querySelectorAll('img').length).not.toBe(0);
  });

  it('should fallback on initials when image is nog available', () => {
    const { getByText, container } = render(<Avatar text="Test Value" />);
    expect(getByText('TV')).toBeInTheDocument();
    expect(container.querySelectorAll('img').length).toBe(0);
  });

  it('should render text conditionally', () => {
    const { getAllByText } = render(
      <>
        <Avatar showLabel text="test-value" />
        <Avatar text="test-value" />
      </>,
    );
    expect(getAllByText('test-value').length).toBe(1);
  });

  it('should render subText', () => {
    const { getAllByText } = render(
      <>
        <Avatar showLabel text="test-value" subText="test-value" />
        <Avatar showLabel text="test-value" />
      </>,
    );
    expect(getAllByText('test-value').length).toBe(3);
  });

  it('should highlight the current user', () => {
    const { container } = render(
      <>
        <Avatar showLabel text="test-value" isCurrentUser />
        <Avatar showLabel text="test-value" />
      </>,
    );

    expect(container).toMatchSnapshot();
  });
});
