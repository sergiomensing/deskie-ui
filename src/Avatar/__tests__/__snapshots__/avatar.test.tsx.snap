// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar should highlight the current user 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 > *:first-child {
  width: 3rem;
  height: 3rem;
  border-radius: 1rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c0 .avatar-initials {
  color: #F7F7FD;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  background-color: #6C5DD3 !important;
}

.c0 .avatar-label {
  margin-left: 0.75rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c0 .avatar-text {
  font-size: 0.875rem;
  color: #6C5DD3;
  font-weight: 500;
}

.c0 .avatar-subtext {
  font-size: 0.8125rem;
  color: #9395A1;
  margin-top: 0.25rem;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 > *:first-child {
  width: 3rem;
  height: 3rem;
  border-radius: 1rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c1 .avatar-initials {
  color: #F7F7FD;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
}

.c1 .avatar-label {
  margin-left: 0.75rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c1 .avatar-text {
  font-size: 0.875rem;
  color: #11142D;
  font-weight: 500;
}

.c1 .avatar-subtext {
  font-size: 0.8125rem;
  color: #9395A1;
  margin-top: 0.25rem;
}

<div>
  <div
    class="c0 avatar "
  >
    <div
      class="avatar-initials"
      style="background-color: rgb(130, 201, 30);"
    >
      tv
    </div>
    <div
      class="avatar-label"
    >
      <span
        class="avatar-text"
      >
        test-value
      </span>
    </div>
  </div>
  <div
    class="c1 avatar "
  >
    <div
      class="avatar-initials"
      style="background-color: rgb(130, 201, 30);"
    >
      tv
    </div>
    <div
      class="avatar-label"
    >
      <span
        class="avatar-text"
      >
        test-value
      </span>
    </div>
  </div>
</div>
`;
